lib/remote_node.go

```diff
@@ -0,0 +1,599 @@
+	ver.Services = SFFullNodeDeprecated
+	if rn.hyperSync {
+		ver.Services |= SFHyperSync
+	}
+	if rn.archivalMode {
+		ver.Services |= SFArchivalNode
+	}
+	if rn.posValidator {
+		ver.Services |= SFPosValidator
+	}
+
+	// We use an int64 instead of a uint64 for convenience.
+	ver.TstampSecs = time.Now().Unix()
+
+	ver.Nonce = nonce
+	ver.UserAgent = rn.params.UserAgent
+
+	// When a node asks you for what height you have, you should reply with the height of the latest actual block you
+	// have. This makes it so that peers who have up-to-date headers but missing blocks won't be considered for initial
+	// block download.
+	ver.StartBlockHeight = uint32(rn.bc.BlockTip().Header.Height)
+
+	// Set the minimum fee rate the peer will accept.
+	ver.MinFeeRateNanosPerKB = rn.minTxFeeRateNanosPerKB
+
+	return ver
+}
+
+// HandleVersionMessage is called upon receiving a version message from the RemoteNode's peer. The peer may be the one
+// initiating the handshake, in which case, we should respond with our own version message. To do this, we pass the
+// responseNonce to this function, which we will use in our response version message.
+func (rn *RemoteNode) HandleVersionMessage(verMsg *MsgDeSoVersion, responseNonce uint64) error {
+	if rn.connectionStatus != RemoteNodeStatus_Connected {
+		return fmt.Errorf("HandleVersionMessage: RemoteNode is not connected")
+	}
+
+	// Verify that the peer's version matches our minimal supported version.
+	if verMsg.Version < rn.params.MinProtocolVersion {
+		return fmt.Errorf("RemoteNode.HandleVersionMessage: Requesting disconnect for id: (%v) "+
+			"protocol version too low. Peer version: %v, min version: %v", rn.id, verMsg.Version, rn.params.MinProtocolVersion)
+		//rn.Disconnect()
+	}
+
+	// Verify that the peer's version message is sent within the version negotiation timeout.
+	if rn.versionTimeExpected.Before(time.Now()) {
+		return fmt.Errorf("RemoteNode.HandleVersionMessage: Requesting disconnect for id: (%v) "+
+			"version timeout. Time expected: %v, now: %v", rn.id, rn.versionTimeExpected.UnixMicro(), time.Now().UnixMicro())
+	}
+
+	vMeta := rn.getHandshakeMetadata()
+	// Record the version the peer is using.
+	vMeta.advertisedProtocolVersion = NewProtocolVersionType(verMsg.Version)
+	// Decide on the protocol version to use for this connection.
+	negotiatedVersion := rn.params.ProtocolVersion
+	if verMsg.Version < rn.params.ProtocolVersion.ToUint64() {
+		negotiatedVersion = NewProtocolVersionType(verMsg.Version)
+	}
+	vMeta.negotiatedProtocolVersion = negotiatedVersion
+
+	// Record the services the peer is advertising.
+	vMeta.ServiceFlag = verMsg.Services
+
+	// Record the tstamp sent by the peer and calculate the time offset.
+	timeConnected := time.Unix(verMsg.TstampSecs, 0)
+	vMeta.timeConnected = &timeConnected
+	currentTime := time.Now().Unix()
+	if currentTime > verMsg.TstampSecs {
+		vMeta.timeOffsetSecs = currentTime - verMsg.TstampSecs
```

Comment:
timeOffsetSecs is an int64, so why not allow it to represent both positive and negative values? This way we know who is "ahead" or "behind". If we do not need this information, then maybe `timeOffsetSecs` should be a uint64.